### PR TITLE
Implement precompile mock contracts and foundry mock util test contracts

### PR DESCRIPTION
**Pull request by mshakeg**
**Description**:

This PR integrates foundry into this hardhat project using the official [hardhat](https://hardhat.org/hardhat-runner/docs/advanced/hardhat-and-foundry) and [foundry](https://book.getfoundry.sh/config/hardhat#use-foundry-in-an-existing-hardhat-project) documentation guides.

It also implements hedera precompile mock contracts and corresponding foundry util test contracts for easy integration and testing in foundry projects.

**Related issue(s)**:

Addresses: https://github.com/hashgraph/hedera-json-rpc-relay/issues/1172

**Notes for reviewer**:

As soon as this merged in, the [foundry-example](https://github.com/hashgraph/hedera-json-rpc-relay/pull/1198) project can install `hedera-smart-contracts` as its only dependency and much of the common code in [this PR](https://github.com/hashgraph/hedera-json-rpc-relay/pull/1198) can be removed, making the `foundry-example` project much simpler.

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: src/Counter.sol [hunk 1]
```diff
@@ -0,0 +1,14 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.9;
+
+contract Counter {
```
Comment (Nana-EC) [inline src/Counter.sol:4]: Move everything under `/src` to under the `/test` or `/contracts/mocks` as appropriate.
We'd like to maintain the `/contracts` folder as a collection of system contracts and reference-able base implementations and `/tests` can have the expansion of edge cases.
Similar to the OZ github

File: script/Counter.s.sol [hunk 2]
```diff
@@ -0,0 +1,12 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.9;
+
+import "forge-std/Script.sol";
+
+contract CounterScript is Script {
```
Comment (Nana-EC) [inline script/Counter.s.sol:6]: This should be under the `/test` folder

**Review by Nana-EC — CHANGES_REQUESTED** [review]
Nice.
Appreciate the links to hardhat and foundry setups - it gave context to why some changes were made.

Content looks good just suggestions on locations to make it fall in line with the folder structure we're going for with the project

File: script/Counter.s.sol (hunk 2 above)
Comment (mshakeg) [reply script/Counter.s.sol:6]: Foundry by default uses the `/script` folder for [solidity scripting](https://book.getfoundry.sh/tutorials/solidity-scripting) hence the `.s.sol` extension and not `.t.sol`, so I don't think it would make sense to move it into `./test/foundry`.

File: contracts/utils/CommonUtils.sol [hunk 3]
```diff
@@ -0,0 +1,43 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.9;
+
+import 'forge-std/Test.sol';
+
+/// generic test utils
+abstract contract CommonUtils is Test {
```
[the remark this answers is not in the document]
Comment (mshakeg) [reply contracts/utils/CommonUtils.sol:7]: Hmm, I initially had the `utils` folder under `./test`, however, moved it into `./contracts` as they'll also be used by developers. Do you still think I should move it into `./test`? It should still work, the base path will just be different.

[the remark this answers is not in the document]
Comment (Nana-EC) [reply contracts/utils/CommonUtils.sol:7]: `./contract` should only contain 
- system contracts interfaces e.g. HTS, exchange and prng right now
- simple reference implementations of the above
- system contract facades e.g. HIP 719
- safe versions of the above as references
- EVM precompiles
- standard contracts e.g. ERC 20, 721 etc

All test contracts and edge cases should be under `./test` as they would have less use in standard contracts and are more for testing or for particular DApp cases that may not relate to many.

That's the plan we're going for, sometime it gets murky though.

File: script/Counter.s.sol (hunk 2 above)
Comment (Nana-EC) [reply script/Counter.s.sol:6]: Okay that's fine. `./scripts` folder makes sense.
Also looking at OZ's repo and seeing a similar approach.
Thanks

Comment (mshakeg) [reply script/Counter.s.sol:6]: OZ's `./scripts` folder doesn't seem to have any solidity scripts, only [hardhat js scripts](https://hardhat.org/hardhat-runner/docs/advanced/scripts)

File: contracts/utils/CommonUtils.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (mshakeg) [reply contracts/utils/CommonUtils.sol:7]: @Nana-EC hmm, ok, though I do think the utils contracts will be commonly used in foundry projects.

In any case, moving `utils` into `./test/foundry` results in the following `hardhat compile` error:
```
Error HH700: Artifact for contract "lib/forge-std/src/Test.sol:Test" not found.
```

**Comment by mshakeg** [comment]
@Nana-EC should I add `artifacts/` to the `.gitignore`? It's typical for the folder to be ignored in hardhat projects and I have many changes to files in that folder and am unsure whether not ignoring that folder was deliberate.

File: contracts/utils/CommonUtils.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (Nana-EC) [reply contracts/utils/CommonUtils.sol:7]: Strange, how does the OZ repo deal with this as we'd like to mimic good practices and organization there where possible?

If `/utils` must remain as is that's fine for this PR, we can figure it out later if compilation breaks on moving

**Comment by Nana-EC** [comment]
> @Nana-EC should I add `artifacts/` to the `.gitignore`? It's typical for the folder to be ignored in hardhat projects and I have many changes to files in that folder and am unsure whether not ignoring that folder was deliberate.

No, not the whole folder.
In the `.gitignore` you'll see we ignore the `build-info` and `dbg.json` files under artifacts.
You can commit the `.sol` files that show up under `artifacts/contracts/`

File: contracts/utils/CommonUtils.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (mshakeg) [reply contracts/utils/CommonUtils.sol:7]: @Nana-EC not sure how OZ got it to work.

I'll try add `./test/foundry` to the [hardhat paths.sources config](https://hardhat.org/hardhat-runner/docs/config#path-configuration).

[the remark this answers is not in the document]
Comment (mshakeg) [reply contracts/utils/CommonUtils.sol:7]: It didn't work :(

**Comment by mshakeg** [comment]
@Nana-EC I'm not sure if Github notified you to [this comment](https://github.com/mshakeg/hedera-smart-contracts/pull/4#discussion_r1199138043) as you're not a collaborator on my fork?

File: contracts/utils/CommonUtils.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (Nana-EC) [reply contracts/utils/CommonUtils.sol]: Okay, we tried best effort. You can leave `/utils` for now and we'll figure it out later

**Comment by mshakeg** [comment]
> Thanks for tagging me, yeah I didn't see the notification but have now responded.
> 
> Btw I would stay simple and iteratively improve as needed since this PR is getting large.

@Nana-EC I replied to your reply, not sure if GitHub notified you now that you're active on that PR. I've had some spare time lately :) I'll try to limit any further work to the tests.

**Comment by Nana-EC** [comment]
> Hey @Nana-EC this PR has been open for quite a while and I might not have as much time to spare in the coming weeks so I'd appreciate it if at least some of the reviewers assigned could spare some time before then.

Hey @mshakeg, yep I've periodically checked. Since the CI tasks were failing I was waiting for them to pass first.
Sure, will push to get reviewers looking at this in the next week or 2.

Thanks again for this.

**Comment by mshakeg** [comment]
Hey @Nana-EC thanks. I did make a few changes that fixed some of the workflow issues pertaining to foundry, however, there's another issue; hardhat seems to be using a gas limit of `21_000` instead of the higher default of ~300k(if gas estimation fails) which is what hardhat uses locally(and hence the test cases pass), so I specified `gas` in the network config in the hopes that it would fix this, however, the issue persists as can be seen [here](https://github.com/mshakeg/hedera-smart-contracts/actions/runs/5225426091/jobs/9434872437). I don't yet know what's causing the discrepancy in local execution and Github actions.

I could try to specify a `gasLimit` for every call and transaction, however, this is non-ideal.

**Comment by Nana-EC** [comment]
> Hey @Nana-EC thanks. I did make a few changes that fixed some of the workflow issues pertaining to foundry, however, there's another issue; hardhat seems to be using a gas limit of `21_000` instead of the higher default of ~300k(if gas estimation fails) which is what hardhat uses locally(and hence the test cases pass), so I specified `gas` in the network config in the hopes that it would fix this, however, the issue persists as can be seen [here](https://github.com/mshakeg/hedera-smart-contracts/actions/runs/5225426091/jobs/9434872437). I don't yet know what's causing the discrepancy in local execution and Github actions.
> 
> I could try to specify a `gasLimit` for every call and transaction, however, this is non-ideal.

By the way the team is making progress on the gasEstimation side.
I believe all function not using HTS system contract methods should work. In the current Mirror Node sprint 0.84 we should begin to see some supported HTS system contract functions.
If you were blocked on some of these then they should begin to function in the next few releases of Mirror Node as HIP 584 work progresses.

**Review by a-ridley — APPROVED** [review]
This looks good to me from my pov. I was able to run the test successfully with no issues.

**Comment by mshakeg** [comment]
Even on hedera-local@2.11.1 hardhat's `auto` gas estimation does not work and it uses a fixed 21_000 for gas which is always inadequate.

The hardhat `gas` network level config override also does not work, possible due to NomicFoundation/hardhat/issues/1228

File: contracts/utils/CommonUtils.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (mshakeg) [reply contracts/utils/CommonUtils.sol]: @Nana-EC what I could do is move the HTS system contract mocks into `./test/foundry` with `/utils`, that way hardhat won't try to compile the mocks which make use of the `./utils` contract? And foundry projects that want to make use of the mocks can import them from `hedera-smart-contracts/test/foundry/...`

File: .github/workflows/test-workflow.yml [hunk 4]
```diff
@@ -8,10 +8,10 @@ on:
         type: string
 
 env:
-  RELAY_IMAGE_TAG: 0.23.3
-  NETWORK_NODE_IMAGE_TAG: 0.39.0-alpha.0
-  HAVEGED_IMAGE_TAG: 0.39.0-alpha.0
-  MIRROR_IMAGE_TAG: 0.81.0-rc1
+  RELAY_IMAGE_TAG: 0.30.0-rc3
```
Comment (mshakeg) [inline .github/workflows/test-workflow.yml:11]: Consistent with the versions used on local-node@latest

File: artifacts/contracts/mocks/exchange-rate-precompile/ExchangeRatePrecompileMock.sol/ExchangeRatePrecompileMock.json [hunk 5]
```diff
@@ -0,0 +1,75 @@
+{
```
Comment (mshakeg) [inline artifacts/contracts/mocks/exchange-rate-precompile/ExchangeRatePrecompileMock.sol/ExchangeRatePrecompileMock.json:1]: If the system contract mocks are all moved into `./test/foundry` along with the `/utils` then these artifacts won't even be generated by hardhat.

**Review by mshakeg — COMMENTED** [review]
@Nana-EC finally had some time to make a few updates and get the test ci in a better state. However it still seems like quite a few test cases fail for the following reasons:

1. if a test case tests a revert, hardhat makes an `eth_estimateGas` request to the hedera local node which returns the expected revert message, however hardhat doesn't match this revert message or attempt the tx(with an overestimated gas) to get the revert message to match
2. contracts intermittently fail to deploy, you can confirm this by comparing the jobs in [run 6133123526](https://github.com/mshakeg/hedera-smart-contracts/actions/runs/6133123526) and [run 6133095468](https://github.com/mshakeg/hedera-smart-contracts/actions/runs/6133095468)

**Comment by Nana-EC** [comment]
> @Nana-EC finally had some time to make a few updates and get the test ci in a better state. However it still seems like quite a few test cases fail for the following reasons:
> 
> 1. if a test case tests a revert, hardhat makes an `eth_estimateGas` request to the hedera local node which returns the expected revert message, however hardhat doesn't match this revert message or attempt the tx(with an overestimated gas) to get the revert message to match
> 2. contracts intermittently fail to deploy, you can confirm this by comparing the jobs in [run 6133123526](https://github.com/mshakeg/hedera-smart-contracts/actions/runs/6133123526) and [run 6133095468](https://github.com/mshakeg/hedera-smart-contracts/actions/runs/6133095468)

Hey @mshakeg.
I saw your ping, I can review this this week.
A couple of products have had version bumps including new features such as mirror node and local node so FYI if you want to bump to take this improvements.
Additional note is the relay has a fix to manage the reverts from Mirror Node as 200's instead of 400 HTTP codes to be consistent with other relay products. If this affects foundry then you might want to wait for the next relay version of 0.33 next week.

I'm hoping we can get simplified working version in soon and improve so let's coordinate on this.
If we can break any of this work out into separate smaller tickets feel free to do that and simplify the PR.

Thanks

**Comment by mshakeg** [comment]
@Nana-EC thanks, will hopefully allocate some time this weekend.

I just upgraded to the latest hedera local node(i.e. `2.14.0`) and the revert tests cases issue as described in (1.) is no longer an issue.

However, a bunch more test cases fail such as the following in `test/erc-20-extensions/ERC20Extensions.js`. This is really strange since I'm running the tests on `hashgraph/hedera-smart-contracts@main:latest` so I would expect them all to pass. Curious to know if this is also an issue on your machine.

How would you want to break up this PR? It's already quite minimal imo. The mocks and tests implemented are relatively basic at this point.

```
  2) ERC20ExtensionsMock tests
       ERC20Burnable tests
         should be able to execute burn(amount):
     Error: Failed to get a different supply value after undefined tries
      at pollForERC20BurnableChangedSupply (utils/helpers.js:75:11)
      at Context.<anonymous> (test/erc-20-extensions/ERC20Extensions.js:93:31)
      at processTicksAndRejections (node:internal/process/task_queues:95:5)

  3) ERC20ExtensionsMock tests
       ERC20Burnable tests
         should be able to execute burnFrom(address, amount):

      AssertionError: expected 800 to equal 900. The numerical values of the given "ethers.BigNumber" and "ethers.BigNumber" inputs were compared, and they differed.
      + expected - actual

      -800
      +900
      
      at Context.<anonymous> (test/erc-20-extensions/ERC20Extensions.js:119:29)
      at processTicksAndRejections (node:internal/process/task_queues:95:5)

  4) ERC20ExtensionsMock tests
       ERC20Pause tests
         should pause and unpause the token:

      AssertionError: expected false to be true
      + expected - actual

      -false
      +true
      
      at Context.<anonymous> (test/erc-20-extensions/ERC20Extensions.js:173:54)
      at processTicksAndRejections (node:internal/process/task_queues:95:5)
```

File: contracts/Counter.sol [hunk 6]
```diff
@@ -0,0 +1,14 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.9;
+
+contract Counter {
```
Comment (Nana-EC) [inline contracts/Counter.sol:4]: Q: what's the value of this contract?
For this repo to avoid it getting humongous we want contracts to either be showcasing Hedera functionality or EVM functionality. 
Goal is similar to the OZ repo, a showcase of functionality and paradigms.
This contract doesn't really do anything special. Is it possible to reuse any of the existing contracts?

If we merge this we'll likely move this later or replace it

File: contracts/mocks/hts-precompile/HtsPrecompileMock.sol [hunk 7]
```diff
@@ -0,0 +1,1839 @@
...
+import '../../../contracts/hts-precompile/KeyHelper.sol';
+import './HederaFungibleToken.sol';
+import './HederaNonFungibleToken.sol';
+import '../../../contracts/NoDelegateCall.sol';
+import '../../../contracts/libraries/Constants.sol';
+
+import '../interfaces/IHtsPrecompileMock.sol';
+import '../libraries/Validation.sol';
+
+contract HtsPrecompileMock is NoDelegateCall, KeyHelper, IHtsPrecompileMock {
```
Comment (Nana-EC) [inline contracts/mocks/hts-precompile/HtsPrecompileMock.sol:16]: nit: We're trying to clarify precompile vs system contract.
If you get a chance
```suggestion
contract HtsSystemContractMock is NoDelegateCall, KeyHelper, IHtsPrecompileMock {
```

File: contracts/mocks/libraries/Validation.sol [hunk 8]
```diff
@@ -0,0 +1,333 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.9;
+
+import '../../../contracts/hts-precompile/HederaResponseCodes.sol';
+import '../hts-precompile/HederaFungibleToken.sol';
+import '../hts-precompile/HederaNonFungibleToken.sol';
+import '../interfaces/IHtsPrecompileMock.sol';
+
+library Validation {
```
Comment (Nana-EC) [inline contracts/mocks/libraries/Validation.sol:9]: Name should inform what is being validated?
```suggestion
library HederaTokenValidation {
```

File: contracts/utils/CommonUtils.sol [hunk 9]
```diff
@@ -0,0 +1,49 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.9;
+
+import 'forge-std/Test.sol';
+
+import '../../contracts/hts-precompile/KeyHelper.sol';
+
+/// generic test utils
+abstract contract CommonUtils is Test, KeyHelper {
```
Comment (Nana-EC) [inline contracts/utils/CommonUtils.sol:9]: Q: should everything under `contracts/utils` be under `contracts/mocks/utils`?
That seems to be where it's used and I'm not clear it's providing value to the example contracts.
They seem all focused on test scenarios as they use or import `forge-std/Test.sol`.

**Review by Nana-EC — CHANGES_REQUESTED** [review]
@mshakeg recent review.
Setup is mostly good

Following items need address
- package file conflicts
- contracts under `/contracts` need to be in an appropriate folder
- `contracts/utils` should probably move to `contracts/mocks/utils` due to it's test focus nature

Everything else seems good

**Comment by Nana-EC** [comment]
> @Nana-EC thanks, will hopefully allocate some time this weekend.
> 
> I just upgraded to the latest hedera local node(i.e. `2.14.0`) and the revert tests cases issue as described in (1.) is no longer an issue.
> 
> However, a bunch more test cases fail such as the following in `test/erc-20-extensions/ERC20Extensions.js`. This is really strange since I'm running the tests on `hashgraph/hedera-smart-contracts@main:latest` so I would expect them all to pass. Curious to know if this is also an issue on your machine.
> 
> How would you want to break up this PR? It's already quite minimal imo. The mocks and tests implemented are relatively basic at this point.
> 
> ```
>   2) ERC20ExtensionsMock tests
>        ERC20Burnable tests
>          should be able to execute burn(amount):
>      Error: Failed to get a different supply value after undefined tries
>       at pollForERC20BurnableChangedSupply (utils/helpers.js:75:11)
>       at Context.<anonymous> (test/erc-20-extensions/ERC20Extensions.js:93:31)
>       at processTicksAndRejections (node:internal/process/task_queues:95:5)
> 
>   3) ERC20ExtensionsMock tests
>        ERC20Burnable tests
>          should be able to execute burnFrom(address, amount):
> 
>       AssertionError: expected 800 to equal 900. The numerical values of the given "ethers.BigNumber" and "ethers.BigNumber" inputs were compared, and they differed.
>       + expected - actual
> 
>       -800
>       +900
>       
>       at Context.<anonymous> (test/erc-20-extensions/ERC20Extensions.js:119:29)
>       at processTicksAndRejections (node:internal/process/task_queues:95:5)
> 
>   4) ERC20ExtensionsMock tests
>        ERC20Pause tests
>          should pause and unpause the token:
> 
>       AssertionError: expected false to be true
>       + expected - actual
> 
>       -false
>       +true
>       
>       at Context.<anonymous> (test/erc-20-extensions/ERC20Extensions.js:173:54)
>       at processTicksAndRejections (node:internal/process/task_queues:95:5)
> ```

Just did a quick review.
The breakout note was if you wanted to. I'm not requiring you break out the PR.
Although review would have definitely been faster if you had broken it out with something like the below. 

1. Foundry config files with a single simple files
2. Initial set of mock contracts for simple HTS tests
3. More complex mock contracts and tests for more detailed tests

Tests from latest were passing as last recent PRs. We'll confirm

File: contracts/Counter.sol (hunk 6 above)
Comment (mshakeg) [reply contracts/Counter.sol:4]: Agreed, this and its corresponding test contract can be removed

File: contracts/NoDelegateCall.sol [hunk 10]
```diff
@@ -0,0 +1,31 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity >=0.8.0;
+
+import "./libraries/Constants.sol";
+
+/// @title Prevents delegatecall to a contract
+/// @notice Base contract that provides a modifier for preventing delegatecall to methods in a child contract
+abstract contract NoDelegateCall is Constants {
```
[the remark this answers is not in the document]
Comment (mshakeg) [reply contracts/NoDelegateCall.sol:8]: This contract is used by the HTS mock contract to enforce `noDelegateCall` on its functions.

File: contracts/utils/CommonUtils.sol (hunk 9 above)
Comment (mshakeg) [reply contracts/utils/CommonUtils.sol:9]: [I previously suggested](https://github.com/hashgraph/hedera-smart-contracts/pull/236#discussion_r1320654970) moving `contracts/mocks` and `contracts/utils` into `test/foundry`. This is pretty standard in a foundry project, and there shouldn't be any hardhat compile issues since hardhat won't try to compile the mocks which requires `forge-std` and fails since that dependency hasn't been configured in hardhat

Comment (Nana-EC) [reply contracts/utils/CommonUtils.sol:9]: @mshakeg This seems like a good idea.
In your current PR `contracts/mocks` is all foundry focused but I can see a general use case of `contracts/mocks` soon.
If I missed this before my bad but yeah let's make this move to make it more self contained for foundry.
Thanks

File: artifacts/contracts/mocks/exchange-rate-precompile/ExchangeRatePrecompileMock.sol/ExchangeRatePrecompileMock.json (hunk 5 above)
Comment (Nana-EC) [reply artifacts/contracts/mocks/exchange-rate-precompile/ExchangeRatePrecompileMock.sol/ExchangeRatePrecompileMock.json:1]: Let's do it

File: contracts/utils/CommonUtils.sol (hunk 9 above)
Comment (mshakeg) [reply contracts/utils/CommonUtils.sol]: Hey @Nana-EC I think I've addressed all of your feedback in this review, also got the tests ci to pass, I just had to re-run all failed jobs a few times, not sure why this is, could be an issue with the local node intermittently not returning the latest and correct state.

File: contracts/NoDelegateCall.sol (hunk 10 above)
[the remark this answers is not in the document]
Comment (Nana-EC) [reply contracts/NoDelegateCall.sol:8]: That's fine. Still think it should be under a folder and not the root `/contracts`. Would it work `/contracts/libraries`?
You can address it here or we'll probably move it later on

**Review by Nana-EC — COMMENTED** [review]
LG
If you can resolve the DCO failure, any other failing workflow and potentially move the `NoDelegateCall.sol` into a folder that would be great.
Ready to sign off after this.
Thanks for the contributions

File: contracts/utils/CommonUtils.sol (hunk 9 above)
Comment (Nana-EC) [reply contracts/utils/CommonUtils.sol]: Thanks @mshakeg 
We've seen a little flakiness with the CI flow but are resolving that.
I just gave a latest review, i think just one more commit and we should be good to merge this.
Thanks for your patience and contributions

**Comment by mshakeg** [comment]
Replaced by https://github.com/hashgraph/hedera-smart-contracts/pull/462
